modules/aac/encode.c

```diff
@@ -30,117 +26,339 @@ static void destructor(void *arg)
 }
 
 
+static void hdr_encode(uint8_t *p, uint16_t size)
+{
+	uint8_t n = 1; /* we only have a single AU-header!!! */
+	/* +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- .. -+-+-+-+-+-+-+-+-+-+
+	 * |AU-headers-length|AU-header|AU-header|      |AU-header|padding|
+	 * |                 |   (1)   |   (2)   |      |   (n)   | bits  |
+	 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+- .. -+-+-+-+-+-+-+-+-+-+
+	 */
+
+	/* AU-headers-length */
+	*(uint16_t *)(void *)&p[0] = htons(sizeof(uint16_t) * 8 * n);
+
+	/* +---------------------------------------+
+	 * |     AU-size                           |
+	 * +---------------------------------------+
+	 * |     AU-Index / AU-Index-delta         |
+	 * +---------------------------------------+
+	 * |     CTS-flag                          |
+	 * +---------------------------------------+
+	 * |     CTS-delta                         |
+	 * +---------------------------------------+
+	 * |     DTS-flag                          |
+	 * +---------------------------------------+
+	 * |     DTS-delta                         |
+	 * +---------------------------------------+
+	 * |     RAP-flag                          |
+	 * +---------------------------------------+
+	 * |     Stream-state                      |
+	 * +---------------------------------------+
+	 */
+
+	/* The AU-header, no CTS, DTS, RAP, Stream-state
+	 *
+	 * AU-size is always the total size of the AU, not the fragmented
+	 * size
+	 */
+	*(uint16_t *)(void *)&p[2] =
+	    htons(size << ((sizeof(uint16_t) * 8) - SIZELENGTH));
+}
+
+
 int aac_encode_update(struct auenc_state **aesp, const struct aucodec *ac,
-		      struct auenc_param *param, const char *fmtp)
+                      struct auenc_param *param, const char *fmtp)
 {
 	struct auenc_state *aes;
+	struct aac_param prm;
+	uint32_t enc_aot, enc_ratio;
+	AACENC_InfoStruct enc_info;
 	AACENC_ERROR error;
 	int err = 0;
-
 	(void)param;
 	(void)fmtp;
 
 	if (!aesp || !ac || !ac->ch)
 		return EINVAL;
 
+	debug("aac: encoder fmtp (%s)\n", fmtp);
+
+	if (str_isset(fmtp)) {
+		aac_mirror_params(fmtp);
+		aac_decode_fmtp(&prm, fmtp);
+	}
+	else {
+		prm.profile_level_id = aac_profile;
+		prm.bitrate = aac_bitrate;
+		prm.constantduration = aac_constantduration;
+	}
+
 	aes = *aesp;
 
-	if (aes)
-		goto out;
+	if (!aes) {
+		aes = mem_zalloc(sizeof(*aes), destructor);
+		if (!aes)
+			return ENOMEM;
 
-	aes = mem_zalloc(sizeof(*aes), destructor);
-	if (!aes)
-		return ENOMEM;
+		error = aacEncOpen(&aes->enc, 0, 0);
```

Comment:
last parameter is "max channels".

Reply:
Some encoding modes fail when parameter is not set to zero. Found it after hours of testing.